File: src/python/fods/Compat/fods_worksheet.py
"""FodsWorksheet — Production facade for a FODS worksheet (table:table).

Spec authority: table:table (FACT-FODS-004, ODF 1.3 §9.1)
TC-W1-FODS-PY-002: Aspose-style deep navigation (FodsWorksheetCollection → FodsWorksheet → cells).
"""
from __future__ import annotations

from typing import Any, ClassVar

from .fods_cell_collection import FodsCellCollection


class FodsWorksheet:
    """Navigable worksheet facade.

    Provides:
      - name        → str
      - rows        → list of raw row dicts
      - cells       → FodsCellCollection for (row, col) navigation
      - is_visible  → bool
    """

    spec_qname: ClassVar[str] = "table:table"
    spec_fact_ref: ClassVar[str] = "FACT-FODS-004"

    def __init__(self, data: dict[str, Any]):
        self._data = data

    @property
    def name(self) -> str:
        """Return the worksheet name (table:name attribute)."""
        return self._data.get("name", "")

    @name.setter
    def name(self, value: str) -> None:
        """Set the worksheet name."""
        self._data["name"] = value

    @property
    def rows(self) -> list[dict[str, Any]]:
        """Return all table:table-row dicts for this worksheet."""
        return self._data.get("rows", [])

    @property
    def cells(self) -> FodsCellCollection:
        """Return the navigable cell collection for (row, col) access."""
        return FodsCellCollection(self.rows)

    @property
    def is_visible(self) -> bool:
        """Return True if the worksheet is visible (table:display != false)."""
        return self._data.get("display", "true").lower() != "false"

    def __repr__(self) -> str:
        return f"FodsWorksheet(name={self.name!r}, rows={len(self.rows)})"

File: src/net/fods/FodsWorksheet.cs
// FormatFactory.Fods — FodsWorksheet
// Public API wrapper for a single ODF spreadsheet sheet (table:table).
// Aspose-style: FodsDocument.Worksheets[i] returns FodsWorksheet.
// Canonical model: FormatFactory.Fods.Model.Table.Table (spec_qname: table:table)
// Authority: plans/.claude/imperative-drifting-conway.md §1, §3
// TC-W1-FODS-NET-003

using System;
using System.Collections.Generic;
using System.Xml.Linq;

namespace FormatFactory.Fods;

/// <summary>
/// Public API wrapper for a single ODF table:table element (a spreadsheet sheet/tab).
/// Backed by the live XElement in the DOM — mutations write through to the document.
///
/// Aspose-style navigation:
///   FodsDocument.Worksheets → FodsWorksheetCollection → FodsWorksheet → FodsRow → FodsCell
///
/// Canonical model: <see cref="FormatFactory.Fods.Model.Table.Table"/>
/// spec_qname: table:table
/// ODF 1.3 §9.1.2
/// </summary>
public sealed class FodsWorksheet
{
    private static readonly XNamespace NsTable =
        "urn:oasis:names:tc:opendocument:xmlns:table:1.0";

    internal XElement Element { get; }

    internal FodsWorksheet(XElement element)
    {
        Element = element;
    }

    // -------------------------------------------------------------------------
    // Name — table:name attribute
    // -------------------------------------------------------------------------

    /// <summary>
    /// Gets or sets the worksheet name (table:name attribute).
    /// ODF 1.3 §9.1.2 — table:name is required.
    /// spec_qname: table:table
    /// </summary>
    public string Name
    {
        get => Element.Attribute(NsTable + "name")?.Value ?? string.Empty;
        set => Element.SetAttributeValue(NsTable + "name", value);
    }

    // -------------------------------------------------------------------------
    // Rows — FodsRowCollection
    // ODF 1.3 §9.4.4: table:table-row children of table:table
    // -------------------------------------------------------------------------

    /// <summary>
    /// All rows in this worksheet (table:table-row elements), in document order.
    /// Each row exposes <see cref="FodsRow.Cells"/> for cell access.
    ///
    /// spec_qname: table:table-row (child QName)
    /// ODF 1.3 §9.4.4
    /// </summary>
    public IReadOnlyList<FodsRow> Rows
    {
        get
        {
            var result = new List<FodsRow>();
            foreach (var child in Element.Elements(NsTable + "table-row"))
                result.Add(new FodsRow(child));
            return result;
        }
    }

    // -------------------------------------------------------------------------
    // Display / Visibility
    // -------------------------------------------------------------------------

    /// <summary>
    /// Gets or sets the sheet visibility (table:display attribute).
    /// True = visible (default); false = hidden.
    /// ODF 1.3 §9.1.2
    /// </summary>
    public bool IsVisible
    {
        get
        {
            var display = Element.Attribute(NsTable + "display")?.Value;
            return display == null || !string.Equals(display, "false",
                StringComparison.OrdinalIgnoreCase);
        }
        set => Element.SetAttributeValue(NsTable + "display", value ? "true" : "false");
    }

    // -------------------------------------------------------------------------
    // FodsSheet compatibility — exposes the underlying XElement via FodsSheet
    // -------------------------------------------------------------------------

    /// <summary>
    /// Returns this worksheet as a <see cref="FodsSheet"/> (legacy DOM wrapper).
    /// Provided for backward compatibility during migration.
    /// Prefer the <see cref="Rows"/> property for new code.
    /// </summary>
    internal FodsSheet AsSheet() => new FodsSheet(Element);
}
